File: ideal-job-scheduler/ideal-job-scheduler-core/src/main/java/com/zzsong/job/scheduler/core/admin/service/JobService.java
package com.zzsong.job.scheduler.core.admin.service;

import com.zzsong.job.common.constants.TriggerTypeEnum;
import com.zzsong.job.common.exception.VisibleException;
import com.zzsong.job.common.transfer.CommonResMsg;
import com.zzsong.job.common.transfer.Paging;
import com.zzsong.job.common.transfer.Res;
import com.zzsong.job.common.utils.DateTimes;
import com.zzsong.job.scheduler.api.dto.req.CreateJobArgs;
import com.zzsong.job.scheduler.api.dto.req.QueryJobArgs;
import com.zzsong.job.scheduler.api.dto.req.UpdateJobArgs;
import com.zzsong.job.scheduler.api.dto.rsp.JobInfoRsp;
import com.zzsong.job.scheduler.core.pojo.JobInfo;
import com.zzsong.job.scheduler.core.pojo.JobView;
import com.zzsong.job.scheduler.core.admin.storage.JobInfoStorage;
import com.zzsong.job.scheduler.core.converter.JobInfoConverter;
import com.zzsong.job.scheduler.core.dispatch.JobDispatcher;
import com.zzsong.job.scheduler.core.utils.CronExpression;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.BeanUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.text.ParseException;
import java.time.LocalDateTime;
import java.util.Date;
import java.util.List;
import java.util.stream.Collectors;

/**
 * @author 宋志宗
 * @date 2020/9/2
 */
@SuppressWarnings("SpringJavaAutowiredFieldsWarningInspection")
@Service
public class JobService {
  private static final Logger log = LoggerFactory.getLogger(JobService.class);
  private static final long PRE_READ_MS = 5000L;

  @Autowired
  private JobDispatcher jobDispatcher;

  private final JobInfoStorage jobInfoStorage;
  private final JobWorkerService jobWorkerService;

  public JobService(JobInfoStorage jobInfoStorage,
                    JobWorkerService jobWorkerService) {
    this.jobInfoStorage = jobInfoStorage;
    this.jobWorkerService = jobWorkerService;
  }

  /**
   * 新建任务
   *
   * @param createJobArgs 新增任务请求参数
   * @return 任务id
   * @author 宋志宗
   * @date 2020/8/26 7:36 下午
   */
  public Mono<JobInfoRsp> createJob(@Nonnull CreateJobArgs createJobArgs) {
    long workerId = createJobArgs.getWorkerId();
    boolean autoStart = createJobArgs.isAutoStart();
    String cron = createJobArgs.getCron();
    return jobWorkerService.loadById(workerId)
        .flatMap(workerOptional -> {
          if (!workerOptional.isPresent()) {
            log.info("新建任务失败, 执行器: {} 不存在", workerId);
            return Mono.error(new VisibleException(CommonResMsg.NOT_FOUND, "执行器不存在"));
          }
          LocalDateTime now = DateTimes.now();
          JobInfo jobInfo = JobInfoConverter.fromCreateJobArgs(createJobArgs);
          jobInfo.setCreatedTime(now);
          jobInfo.setUpdateTime(now);
          if (StringUtils.isNotBlank(cron)) {
            boolean validExpression = CronExpression.isValidExpression(cron);
            if (!validExpression) {
              log.info("cron表达式: {} 不合法", cron);
              return Mono.error(new VisibleException(CommonResMsg.BAD_REQUEST, "cron表达式不合法"));
            }
            if (autoStart) {
              long nextTriggerTime = getNextTriggerTime(cron);
              jobInfo.setJobStatus(JobInfo.JOB_START);
              jobInfo.setNextTriggerTime(nextTriggerTime);
            }
          }
          return jobInfoStorage.save(jobInfo).map(JobInfoConverter::toJobInfoRsp);
        });
  }

  /**
   * 更新任务信息
   *
   * @param args 更新参数
   * @author 宋志宗
   * @date 2020/8/26 8:48 下午
   */
  public Mono<JobInfoRsp> updateJob(@Nonnull UpdateJobArgs args) {
    long jobId = args.getJobId();
    String cron = args.getCron();
    return jobInfoStorage.findById(jobId)
        .flatMap(jobInfoOptional -> {
          if (!jobInfoOptional.isPresent()) {
            log.info("任务: {} 不存在", jobId);
            return Mono.error(new VisibleException(CommonResMsg.NOT_FOUND, "任务不存在"));
          }
          LocalDateTime now = DateTimes.now();
          JobInfo jobInfo = jobInfoOptional.get();
          BeanUtils.copyProperties(args, jobInfo);
          jobInfo.setUpdateTime(now);
          if (StringUtils.isNotBlank(cron)) {
            boolean validExpression = CronExpression.isValidExpression(cron);
            if (!validExpression) {
              log.info("cron表达式: {} 不合法", cron);
              return Mono.error(new VisibleException(CommonResMsg.BAD_REQUEST, "cron表达式不合法"));
            }
            if (jobInfo.getJobStatus() == JobInfo.JOB_START) {
              long nextTriggerTime = getNextTriggerTime(cron);
              jobInfo.setNextTriggerTime(nextTriggerTime);
            }
          } else {
            jobInfo.setJobStatus(JobInfo.JOB_STOP);
          }
          return jobInfoStorage.save(jobInfo).map(JobInfoConverter::toJobInfoRsp);
        });
  }

  /**
   * 移除任务
   *
   * @param jobId 任务id
   * @author 宋志宗
   * @date 2020/8/26 8:49 下午
   */
  public Mono<Integer> removeJob(long jobId) {
    return jobInfoStorage.findById(jobId)
        .flatMap(jobInfoOptional -> {
          if (!jobInfoOptional.isPresent()) {
            log.info("任务: {} 不存在", jobId);
            return Mono.just(0);
          }
          return jobInfoStorage.delete(jobId);
        });
  }

  /**
   * 查询任务信息
   *
   * @param args   查询参数
   * @param paging 分页参数
   * @return 任务信息列表
   * @author 宋志宗
   * @date 2020/8/26 8:51 下午
   */
  @Nonnull
  public Mono<Res<List<JobInfoRsp>>> query(@Nonnull QueryJobArgs args, @Nonnull Paging paging) {
    return jobInfoStorage.query(args, paging)
        .map(listRes ->
            listRes.convertNewRes(list ->
                list.stream()
                    .map(JobInfoConverter::toJobInfoRsp)
                    .collect(Collectors.toList())
            )
        );
  }

  /**
   * 启用任务
   *
   * @param jobId 任务id
   * @author 宋志宗
   * @date 2020/8/20 4:38 下午
   */
  public Mono<Boolean> enableJob(long jobId) {
    return jobInfoStorage.findById(jobId)
        .flatMap(jobInfoOptional -> {
          if (!jobInfoOptional.isPresent()) {
            log.info("任务: {} 不存在", jobId);
            return Mono.error(new VisibleException(CommonResMsg.NOT_FOUND, "任务不存在"));
          }
          JobInfo jobInfo = jobInfoOptional.get();
          if (jobInfo.getJobStatus() == JobInfo.JOB_START) {
            log.info("任务: {} 正在在运行中", jobId);
            return Mono.just(true);
          }
          String cron = jobInfo.getCron();
          if (StringUtils.isBlank(cron)) {
            log.info("启动任务: {} 失败, cron表达式为空", jobId);
            return Mono.error(new VisibleException("cron表达式为空"));
          }
          long nextTriggerTime = getNextTriggerTime(cron);
          jobInfo.setJobStatus(JobInfo.JOB_START);
          jobInfo.setLastTriggerTime(0);
          jobInfo.setNextTriggerTime(nextTriggerTime);
          return jobInfoStorage.save(jobInfo).map(j -> true);
        });
  }

  /**
   * 停用任务
   *
   * @param jobId 任务id
   * @author 宋志宗
   * @date 2020/8/20 4:38 下午
   */
  public Mono<Boolean> disableJob(long jobId) {
    return jobInfoStorage.findById(jobId)
        .flatMap(jobInfoOptional -> {
          if (!jobInfoOptional.isPresent()) {
            log.info("任务: {} 不存在", jobId);
            return Mono.error(new VisibleException(CommonResMsg.NOT_FOUND, "任务不存在"));
          }
          JobInfo jobInfo = jobInfoOptional.get();
          if (jobInfo.getJobStatus() == JobInfo.JOB_STOP) {
            log.info("任务: {} 为停止状态", jobId);
            return Mono.just(true);
          }
          jobInfo.setJobStatus(JobInfo.JOB_STOP);
          jobInfo.setLastTriggerTime(0);
          jobInfo.setNextTriggerTime(0);
          return jobInfoStorage.save(jobInfo).map(j -> true);
        });
  }

  public Mono<Boolean> triggerJob(long jobId, @Nullable String customExecuteParam) {
    return jobInfoStorage.findJobViewById(jobId)
        .flatMap(jobViewOptional -> {
          if (!jobViewOptional.isPresent()) {
            log.info("任务: {} 不存在", jobId);
            return Mono.error(new VisibleException(CommonResMsg.NOT_FOUND, "任务不存在"));
          }
          JobView jobView = jobViewOptional.get();
          return jobDispatcher.dispatch(jobView, TriggerTypeEnum.MANUAL, customExecuteParam);
        });
  }

  /**
   * 读取一批待执行的任务
   *
   * @param maxNextTime 最大下次执行时间
   * @param count       读取数量
   * @return 待执行任务列表
   * @date 2020/8/24 8:46 下午
   */
  public Flux<JobView> loadScheduleJobViews(long maxNextTime, int count) {
    Paging paging = Paging.of(1, count).ascBy("jobId");
    return jobInfoStorage.loadScheduleJobViews(maxNextTime, paging);
  }

  public Mono<Integer> batchUpdateTriggerInfo(List<JobView> viewList) {
    return jobInfoStorage.batchUpdateTriggerInfo(viewList);
  }


  private long getNextTriggerTime(String cron) {
    long benchmark = System.currentTimeMillis() + PRE_READ_MS;
    Date nextValidTime;
    try {
      nextValidTime = new CronExpression(cron)
          .getNextValidTimeAfter(new Date(benchmark));
    } catch (ParseException e) {
      String errMsg = e.getClass().getName() + ": " + e.getMessage();
      log.info("解析cron: {} 异常: {}", cron, errMsg);
      throw new VisibleException("解析cron表达式出现异常");
    }
    if (nextValidTime == null) {
      log.info("尝试通过cron: {} 获取下次执行时间失败", cron);
      throw new VisibleException("获取下次执行时间失败");
    }
    return nextValidTime.getTime();
  }
}

File: ideal-job-scheduler/ideal-job-scheduler-core/src/main/kotlin/cn/sh/ideal/job/scheduler/core/admin/service/JobService.kt
package cn.sh.ideal.job.scheduler.core.admin.service

import cn.sh.ideal.job.common.constants.*
import cn.sh.ideal.job.common.exception.VisibleException
import cn.sh.ideal.job.common.transfer.CommonResMsg
import cn.sh.ideal.job.common.transfer.Paging
import cn.sh.ideal.job.common.transfer.Res
import cn.sh.ideal.job.common.transfer.SpringPages
import cn.sh.ideal.job.scheduler.api.dto.req.CreateJobArgs
import cn.sh.ideal.job.scheduler.api.dto.req.QueryJobArgs
import cn.sh.ideal.job.scheduler.api.dto.req.UpdateJobArgs
import cn.sh.ideal.job.scheduler.api.dto.rsp.JobInfoRsp
import cn.sh.ideal.job.scheduler.core.admin.entity.JobInfo
import cn.sh.ideal.job.scheduler.core.admin.entity.vo.DispatchJobView
import cn.sh.ideal.job.scheduler.core.admin.repository.JobInfoRepository
import cn.sh.ideal.job.scheduler.core.converter.JobInfoConverter
import cn.sh.ideal.job.scheduler.core.dispatch.JobDispatch
import cn.sh.ideal.job.scheduler.core.utils.CronExpression
import org.apache.commons.lang3.StringUtils
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import java.util.*
import javax.persistence.criteria.Predicate

/**
 * @author 宋志宗
 * @date 2020/8/20
 */
@Suppress("DuplicatedCode")
@Service
class JobService(private val jobInfoRepository: JobInfoRepository) {
  private val log: Logger = LoggerFactory.getLogger(this.javaClass)
  private val preReadMs = 5000L

  @Autowired
  private lateinit var jobDispatch: JobDispatch

  @Autowired
  private lateinit var jobExecutorService: JobExecutorService

  /**
   * 新建任务
   *
   * @param createJobArgs 新增任务请求参数
   * @return 任务id
   * @author 宋志宗
   * @date 2020/8/26 7:36 下午
   */
  fun createJob(createJobArgs: CreateJobArgs): Res<Long> {
    val executorId = createJobArgs.executorId
    val autoStart = createJobArgs.autoStart ?: false
    val cron = createJobArgs.cron ?: ""
    val executor = jobExecutorService.loadById(executorId)
    if (executor == null) {
      log.info("新建任务失败, 执行器: {} 不存在", executorId)
      return Res.err(CommonResMsg.NOT_FOUND, "执行器不存在")
    }
    val jobInfo = JobInfoConverter.fromCreateJobArgs(createJobArgs)
    if (autoStart && cron.isNotBlank()) {
      val validExpression = CronExpression.isValidExpression(cron)
      if (!validExpression) {
        return Res.err(CommonResMsg.BAD_REQUEST, "cron表达式不合法")
      }
      val benchmark = System.currentTimeMillis() + preReadMs
      val nextValidTime = try {
        CronExpression(cron).getNextValidTimeAfter(Date(benchmark))
      } catch (exception: Exception) {
        val errMessage = "${exception.javaClass.name}:${exception.message}"
        log.info("解析cron: {} 异常: {}", cron, errMessage)
        return Res.err(CommonResMsg.INTERNAL_SERVER_ERROR, "解析cron表达式出现异常")
      }
      if (nextValidTime == null) {
        log.info("尝试通过cron: {} 获取下次执行时间失败", cron)
        return Res.err(CommonResMsg.BAD_REQUEST, "获取下次执行时间失败")
      }
      jobInfo.jobStatus = JobInfo.JOB_START
      val nextTriggerTime = nextValidTime.time
      jobInfo.nextTriggerTime = nextTriggerTime
    }
    jobInfoRepository.save(jobInfo)
    val jobId = jobInfo.jobId
    return Res.data(jobId)
  }

  /**
   * 更新任务信息
   *
   * @param updateJobArgs 更新参数
   * @return 更新结果
   * @author 宋志宗
   * @date 2020/8/26 8:48 下午
   */
  fun updateJob(updateJobArgs: UpdateJobArgs): Res<Void> {
    val jobId = updateJobArgs.jobId
    val jobInfo = jobInfoRepository.findByIdOrNull(jobId)
    if (jobInfo == null) {
      log.info("任务: {} 不存在", jobId)
      return Res.err(CommonResMsg.NOT_FOUND, "任务不存在")
    }
    val executorId = updateJobArgs.executorId
    val executorHandler = updateJobArgs.executorHandler ?: DBDefaults.DEFAULT_STRING_VALUE
    val executeParam = updateJobArgs.executeParam ?: DBDefaults.DEFAULT_STRING_VALUE
    val routeStrategy = updateJobArgs.routeStrategy ?: RouteStrategyEnum.POLLING
    val blockStrategy = updateJobArgs.blockStrategy ?: BlockStrategyEnum.SERIAL
    val cron = updateJobArgs.cron ?: DBDefaults.DEFAULT_STRING_VALUE
    val retryCount = updateJobArgs.retryCount ?: DBDefaults.DEFAULT_INT_VALUE
    val jobName = updateJobArgs.jobName ?: DBDefaults.DEFAULT_STRING_VALUE
    val alarmEmail = updateJobArgs.alarmEmail ?: DBDefaults.DEFAULT_STRING_VALUE

    jobInfo.executorId = executorId
    jobInfo.executorHandler = executorHandler
    jobInfo.executeParam = executeParam
    jobInfo.routeStrategy = routeStrategy
    jobInfo.blockStrategy = blockStrategy
    jobInfo.retryCount = retryCount
    jobInfo.jobName = jobName
    jobInfo.alarmEmail = alarmEmail
    if (cron.isNotBlank()) {
      val validExpression = CronExpression.isValidExpression(cron)
      if (!validExpression) {
        return Res.err(CommonResMsg.BAD_REQUEST, "cron表达式不合法")
      }
      jobInfo.cron = cron
      if (jobInfo.jobStatus == JobInfo.JOB_START) {
        val benchmark = System.currentTimeMillis() + preReadMs
        val nextValidTime = try {
          CronExpression(cron).getNextValidTimeAfter(Date(benchmark))
        } catch (exception: Exception) {
          val errMessage = "${exception.javaClass.name}:${exception.message}"
          log.info("解析cron: {} 异常: {}", cron, errMessage)
          return Res.err(CommonResMsg.INTERNAL_SERVER_ERROR, "解析cron表达式出现异常")
        }
        if (nextValidTime == null) {
          log.info("尝试通过cron: {} 获取下次执行时间失败", cron)
          return Res.err(CommonResMsg.BAD_REQUEST, "获取下次执行时间失败")
        }
        val nextTriggerTime = nextValidTime.time
        jobInfo.nextTriggerTime = nextTriggerTime
      }
    }
    jobInfoRepository.save(jobInfo)
    return Res.success()
  }

  /**
   * 移除任务
   *
   * @param jobId 任务id
   * @return 移除结果
   * @author 宋志宗
   * @date 2020/8/26 8:49 下午
   */
  fun removeJob(jobId: Long): Res<Void> {
    val jobInfo = jobInfoRepository.findByIdOrNull(jobId)
    if (jobInfo == null) {
      log.info("任务: {} 不存在", jobId)
      return Res.success(CommonResMsg.NOT_FOUND)
    }
    jobInfoRepository.delete(jobInfo)
    return Res.success()
  }

  /**
   * 查询任务信息
   *
   * @param args   查询参数
   * @param paging 分页参数
   * @return 任务信息列表
   * @author 宋志宗
   * @date 2020/8/26 8:51 下午
   */
  fun query(args: QueryJobArgs, paging: Paging): Res<List<JobInfoRsp>> {
    val executorId = args.executorId
    val jobName = args.jobName
    val executorHandler = args.executorHandler
    val jobStatus = args.jobStatus
    val application = args.application
    val tenantId = args.tenantId
    val bizType = args.bizType
    val customTag = args.customTag
    val businessId = args.businessId

    val page = jobInfoRepository.findAll(
        { root, cq, cb ->
          val predicates = ArrayList<Predicate>()
          if (executorId != null && executorId > 0) {
            predicates.add(cb.equal(root.get<Any>("executorId"), executorId))
          }
          if (StringUtils.isNotBlank(jobName)) {
            predicates.add(cb.equal(root.get<Any>("jobName"), jobName))
          }
          if (StringUtils.isNotBlank(executorHandler)) {
            predicates.add(cb.equal(root.get<Any>("executorHandler"), executorHandler))
          }
          if (jobStatus != null && jobStatus == JobInfo.JOB_START) {
            predicates.add(cb.equal(root.get<Any>("jobStatus"), JobInfo.JOB_START))
          } else if (jobStatus != null && jobStatus == JobInfo.JOB_STOP) {
            predicates.add(cb.equal(root.get<Any>("jobStatus"), JobInfo.JOB_STOP))
          }
          if (StringUtils.isNotBlank(application)) {
            predicates.add(cb.equal(root.get<Any>("application"), application))
          }
          if (StringUtils.isNotBlank(tenantId)) {
            predicates.add(cb.equal(root.get<Any>("tenantId"), tenantId))
          }
          if (StringUtils.isNotBlank(bizType)) {
            predicates.add(cb.equal(root.get<Any>("bizType"), bizType))
          }
          if (StringUtils.isNotBlank(customTag)) {
            predicates.add(cb.equal(root.get<Any>("customTag"), customTag))
          }
          if (StringUtils.isNotBlank(businessId)) {
            predicates.add(cb.equal(root.get<Any>("businessId"), businessId))
          }
          cq.where(*predicates.toTypedArray()).restriction
        }, SpringPages.paging2Pageable(paging)
    )
    return SpringPages.toPageRes(page) { JobInfoConverter.toJobInfoRsp(it) }
  }

  /**
   * 启用任务
   *
   * @param jobId 任务id
   * @return 执行结果
   * @author 宋志宗
   * @date 2020/8/20 4:38 下午
   */
  fun enableJob(jobId: Long): Res<Void> {
    val jobInfo = jobInfoRepository.findByIdOrNull(jobId)
    if (jobInfo == null) {
      log.info("任务: {} 不存在", jobId)
      return Res.err(CommonResMsg.NOT_FOUND, "任务不存在")
    }
    if (jobInfo.jobStatus == JobInfo.JOB_START) {
      log.info("任务: {} 正在在运行中", jobId)
      return Res.success()
    }
    val cron = jobInfo.cron
    if (cron.isBlank()) {
      log.info("启动任务: {} 失败, cron表达式为空", jobId)
      return Res.err(CommonResMsg.BAD_REQUEST, "cron表达式不合法")
    }
    val benchmark = System.currentTimeMillis() + preReadMs
    val nextValidTime = try {
      CronExpression(cron).getNextValidTimeAfter(Date(benchmark))
    } catch (exception: Exception) {
      val errMessage = "${exception.javaClass.name}:${exception.message}"
      log.info("解析cron: {} 异常: {}", cron, errMessage)
      return Res.err(CommonResMsg.INTERNAL_SERVER_ERROR, "解析cron表达式出现异常")
    }
    if (nextValidTime == null) {
      log.info("尝试通过cron: {} 获取下次执行时间失败", cron)
      return Res.err(CommonResMsg.BAD_REQUEST, "获取下次执行时间失败")
    }
    jobInfo.jobStatus = JobInfo.JOB_START
    jobInfo.lastTriggerTime = 0
    val nextTriggerTime = nextValidTime.time
    jobInfo.nextTriggerTime = nextTriggerTime
    jobInfoRepository.save(jobInfo)
    return Res.success()
  }

  /**
   * 停用任务
   *
   * @param jobId 任务id
   * @return 执行结果
   * @author 宋志宗
   * @date 2020/8/20 4:38 下午
   */
  fun disableJob(jobId: Long): Res<Void> {
    val jobInfo = jobInfoRepository.findByIdOrNull(jobId)
    if (jobInfo == null) {
      log.info("任务: {} 不存在", jobId)
      return Res.err(CommonResMsg.NOT_FOUND, "任务不存在")
    }
    if (jobInfo.jobStatus == JobInfo.JOB_STOP) {
      log.info("任务: {} 为停止状态", jobId)
      return Res.success()
    }
    jobInfo.jobStatus = JobInfo.JOB_STOP
    jobInfo.lastTriggerTime = 0
    jobInfo.nextTriggerTime = 0
    jobInfoRepository.save(jobInfo)
    return Res.success()
  }

  /**
   * 触发任务
   *
   * @param jobId              任务id
   * @param customExecuteParam 执行参数, 如果为<code>null</code则使用默认任务配置
   * @author 宋志宗
   * @date 2020/8/24 8:46 下午
   */
  fun trigger(jobId: Long, customExecuteParam: String?) {
    val dispatchJobView = jobInfoRepository.findDispatchJobViewById(jobId)
    if (dispatchJobView == null) {
      log.info("任务: {} 不存在", jobId)
      throw VisibleException(CommonResMsg.NOT_FOUND, "任务不存在")
    }
    jobDispatch.dispatch(dispatchJobView, TriggerTypeEnum.MANUAL, customExecuteParam)
  }

  fun existsByExecutorId(executorId: Long): Boolean {
    return jobInfoRepository.existsByExecutorId(executorId)
  }

  fun loadScheduleJobViews(maxNextTime: Long, count: Int): List<DispatchJobView> {
    val sort = Sort.by("jobId").ascending()
    val pageRequest = PageRequest.of(0, count, sort)
    val jobStart = JobInfo.JOB_START
    return jobInfoRepository.loadScheduleJobViews(jobStart, maxNextTime, pageRequest)
  }

  fun batchUpdateTriggerInfo(jobInfoList: List<DispatchJobView>) {
    jobInfoRepository.batchUpdateTriggerInfo(jobInfoList)
  }
}